Governable random beacon parameters

tomaszslabon opened the pull request:
#Depends on https://github.com/keep-network/keep-core/pull/2612
This PR adds a contract containing governable parameters that will be needed by Random Beacon v2.
The parameters are taken from [RFC-19](https://github.com/keep-network/keep-core/pull/2585/files#).


Question: How should the parameters be initialised? One option is to initialize them via a contractor (at least those parameters that take long to initialize).  


The unit tests were executed locally and pass, but they do not execute on CI (other PRs should enable this).
The solidity and typescript code were analysed with tools like `solhint`,`prettier`, `eslint` for typescript (configuration for those mechanisms should probably be added by other PRs).

nkuba commented on solidity/random-beacon/test/helpers/contract-test-helpers.ts:
```diff
@@ -0,0 +1,11 @@
+import { ethers } from "hardhat"
```
What if we try to extract helpers to some common library, e.g. https://github.com/keep-network/hardhat-helpers?

tomaszslabon replied:
@nkuba I'm OK with keeping functions like `increaseTime` in the common library like `hardhat-helpers`.
@pdyraga What do you think?

pdyraga replied:
Yes, we are repeating this code in every repository. It is time to extract it.

tomaszslabon replied:
If you don't mind, we could do it in a separate PR.

nkuba replied:
I did it in https://github.com/keep-network/hardhat-helpers/pull/5

tomaszslabon replied:
I used `hardhat-helpers` in https://github.com/keep-network/keep-core/pull/2613/commits/3506fdfc60ca6aed60af446299610f6fcd69a6f7.
I added the package vie the GitHub url with the latest commit hash:
"@keep-network/hardhat-helpers": "https://github.com/keep-network/hardhat-helpers#ed35ab7",

If we release the new version of `hardhat-helper` then we can replan it with that package instead of using GitHub url.

pdyraga commented on solidity/random-beacon/contracts/RandomBeaconGovernable.sol:
```diff
@@ -0,0 +1,693 @@
+    uint256 public newGroupCreationFrequency;
+    uint256 public groupCreationFrequencyChangeInitiated;
+
+    /// @notice Group lifetime
```
`in seconds`

tomaszslabon replied:
Done in https://github.com/keep-network/keep-core/pull/2613/commits/29f18f6ce0f8b89b14a1f21f8c801aa9fbea1ace.

pdyraga commented on solidity/random-beacon/contracts/RandomBeaconGovernable.sol:
```diff
@@ -0,0 +1,693 @@
+    uint256 public newDkgSubmissionEligibilityDelay;
+    uint256 public dkgSubmissionEligibilityDelayChangeInitiated;
+
+    /// @notice Hard timeout for a relay entry
```
`in blocks`

tomaszslabon replied:
Done in https://github.com/keep-network/keep-core/pull/2613/commits/29f18f6ce0f8b89b14a1f21f8c801aa9fbea1ace.

pdyraga commented:
I see we assume a global governance delay of 7 days for all parameters and I do not think global governance delay works well in this case. The role of governance delay is to protect stakers against harmful changes and give them a chance to opt out in case they do not agree with the proposed change. With a global governance delay of 7 days, it is possible stakers would be forced to work under certain conditions - say group lifetime is 30 days, slashing penalty increases to 1M per staker, and relay entry submission eligibility delay is reduced to 1 block.

One option would be to capture the *current* parameters of the system for each newly created group and to guarantee the rules never change for that particular group staker qualified to. This is pretty similar to how tBTC deposit works. The downside of this approach is that everything would be much more expensive to execute on-chain, and external applications would not have a consistent experience working with the beacon - there is no guarantee which group will be selected to produce a new relay entry and each group might have its own parameters for a timeout or slashing.

Another option would be to require that the governance delay > group lifetime, or even that governance delay = group lifetime. This is similar to the coverage pool's unstake delay. The downside though is that this approach would be quite unwieldy for the random beacon - if the group lifetime is 30 days, do we really want to wait for that long to modify group creation frequency?

Yet one option is to apply range checks for some parameters and set governance delays individually:
- `relayRequestFee`: no validation, governance delay of 24 hours (goes to maintenance pool and DKG cost is predictable)
- `dkgResultSubmissionReward`: no validation, governance delay of 24 hours (DKG cost is predictable)
- `sortitionPoolUnlockingReward`: no validation, governance delay of 24 hours (can not harm stakers)
- `maliciousDkgResultSlashingAmount`: no validation, governance delay of 24 hours (malicious behavior, nothing to protect)
- `dkgResultChallengePeriodLength`: must be > 10 blocks, governance delay of 24 hours (10 blocks should be enough in most cases)
- `relayEntrySubmissionEligibilityDelay`: must be > 0 blocks, governance delay of 24 hours (64 blocks should be enough to submit entry before a timeout)
- `dkgSubmissionEligibilityDelay`: must be > 0 blocks, governance delay of 24 hours (64 blocks should be enough to submit DKG result before a timeout)
- `callbackGasLimit`: must be > 0 and < 1M, governance delay of 2 weeks
- `groupCreationFrequency`: must be > 0, governance delay of 2 weeks
- `relayEntrySubmissionFailureSlashingAmount`: no validation, governance delay of 2 weeks
- `relayEntryHardTimeout`: no validation, governance delay of 2 weeks
- `groupLifetime`: must be >= 1 day and <= 2 weeks

tomaszslabon commented on solidity/random-beacon/contracts/GovernanceUtils.sol:
```diff
@@ -0,0 +1,37 @@
+
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.5;
```
Done in https://github.com/keep-network/keep-core/pull/2613/commits/34e676b9e6949ea44f816e3636d2acf37286c5fc.

tomaszslabon commented on solidity/random-beacon/contracts/RandomBeaconGovernable.sol:
```diff
@@ -0,0 +1,693 @@
+
+/// @title Random Beacon Governable
+/// @notice This is a base contract that holds governable parameters for Random
+///        Beacon V2 as well as their update logic
```
Done in https://github.com/keep-network/keep-core/pull/2613/commits/29f18f6ce0f8b89b14a1f21f8c801aa9fbea1ace.

pdyraga commented:
I am concerned about the contract size. I measured `RandomBeaconGovernable` contract size and it is more than 12KB, while Ethereum limit for contract size is 24KB. It means that just the governance part takes more than 50% of the available space.

After marking all `new*` and `*ChangeInitiated` fields as `internal`, the contract size went down to 10.9 KB, and after removing all `getRemaining*` functions, the contract size went down to 9.29 KB. This is less but still not enough - more than 37% of the allowed size is occupied. This is not a problem right now but will become a problem as soon as we start building the main logic.

I would recommend experimenting with extracting the governance logic to a library and adjusting types of fields. 

In particular, no more than `(unit96).max` of T will ever be minted (see https://github.com/threshold-network/solidity-contracts/pull/8) so we should be able to alter all rewards, amounts, fees to use `uint96` instead of `uint256` and try to pack them densely with timeouts and delays for which `uint32` should be more than enough.

All the logic regarding governance delays and validation would exist in the library, reducing the contract to just a set of `external` functions referencing ones from the library (it is just DELEGATECALL). I would also remove all `getRemaining*` functions - they are nice-to-have but not required for the system to work, everything can be read from events. We do not need to expose `new*` and `*ChangeInitiated` public - this information can also be read from events. Instead of having a separate field for each `*ChangeInitiated` I think we can use `mapping(string => uint256)` indexed by parameter name.

I think we still need to have a parent contract for the governance to keep there all the boilerplate: events need to be re-declared in the contract, we need to have `external` functions referencing ones from the library, and we can do there `uint* -> uint256` conversions for convenience. 

Something like:

```
library GovernableParameters {
    struct Storage {
        uint96 relayRequestFee;
        uint96 relayEntrySubmissionFailureSlashingAmount;
        uint32 relayEntrySubmissionEligibilityDelay;
        uint32 relayEntryHardTimeout;

        uint96 dkgResultSubmissionReward;
        uint96 maliciousDkgResultSlashingAmount;
        uint32 dkgSubmissionEligibilityDelay;
        uint32 dkgResultChallengePeriodLength;

        uint96 sortitionPoolUnlockingReward;
        uint32 groupCreationFrequency;
        uint32 groupLifetime;
        uint32 callbackGasLimit;

        mapping(string => uint256) changeInitiatedTimestamp;
    }

    event RelayRequestFeeUpdateStarted(
        uint256 relayRequestFee,
        uint256 timestamp
    );
    event RelayRequestFeeUpdated(uint256 relayRequestFee);

    // (...)

    modifier onlyAfterGovernanceDelay(
        Storage storage self,
        string memory parameter,
        uint256 delay
    ) { 
        uint256 initiated = self.changeInitiatedTimestamp(parameter);
        // ... 
    }

    function getRemainingChangeTime(uint256 changeTimestamp, uint256 delay)
        internal
        view
        returns (uint256)
    { 
        // ... 
    }

    function beginRelayRequestFeeUpdate(uint256 newRelayRequestFee) external {
        // (...)
    }

    function finalizeRelayRequestFeeUpdate() 
        external 
        onlyAfterGovernanceDelay("relayRequestFee", 24 hours) {
        // (...)
    }

    // (...)
}
```

```
contract RandomBeaconParameters is Ownable {
    using GovernableParameters for GovernableParameters.Storage; 

    event RelayRequestFeeUpdateStarted(
        uint256 relayRequestFee,
        uint256 timestamp
    );
    event RelayRequestFeeUpdated(uint256 relayRequestFee);

    GovernableParameters.Storage governableParameters;

    // (...)

    function beginRelayRequestFeeUpdate(uint256 newRelayRequestFee) external onlyOwner {
        governableParameters.beginRelayRequestFeeUpdate(newRelayRequestFee);
    }

    function finalizeRelayRequestFeeUpdate() external onlyOwner {
        governableParameters.finalizeRelayRequestFeeUpdate();
    }

    // (...)
    
    function relayRequestFee() public {
        return uint256(governableParameters.relayRequestFee);
    }
}
```

tomaszslabon commented:
@pdyraga 
I moved the governable parameters into a library as [suggested by you](https://github.com/keep-network/keep-core/pull/2613#issuecomment-927851406) in https://github.com/keep-network/keep-core/pull/2613/commits/d0e0043dd0c22d4d560145706b2d3052622edc7b.

I added validation of parameters as [suggested by you](https://github.com/keep-network/keep-core/pull/2613#issuecomment-927813679) in https://github.com/keep-network/keep-core/pull/2613/commits/32e422413516f6dd9e88d971bee8f60173c4ff97.

pdyraga reviewed (commented):
I was experimenting with the contract size and I am afraid we still need to do better. The current version is more optimal than the previous one, but we are still above 9KB. I tried various options and the best I could come up with is to split all the governance-related functionalities into a separate contract, called `RandomBeaconGovernance` that would own `RandomBeacon` contract.

`RandomBeaconGovernance` would contain all `initialize*` and `finalize*` functions and we can even put there `getRemaining*` functions we originally had. `RandomBeacon` would expose four functions allowing the governance to update parameters, grouped by the parameter type (relay request, DKG, slashing, rewards). I'd also skip all the complexity of other types than `uint256`.

This should be pretty gas-optimal and the size of `RandomBeacon` contract would be just about 2.5KB. 

What do you think about it? 

Below is a sample code to help to visualize this proposition:
```
pragma solidity ^0.8.6;

import "@openzeppelin/contracts/access/Ownable.sol";

contract RandomBeaconGovernance is Ownable {
    uint256 public newRelayRequestFee;
    uint256 public relayRequestFeeChangeInitiated;

    uint256 public newDkgResultSubmissionReward;
    uint256 public dkgResultSubmissionRewardChangeInitiated;

    uint256 public newSortitionPoolUnlockingReward;
    uint256 public sortitionPoolUnlockingRewardChangeInitiated;

    uint256 public newRelayEntrySubmissionFailureSlashingAmount;
    uint256 public relayEntrySubmissionFailureSlashingAmountChangeInitiated;

    uint256 public newMaliciousDkgResultSlashingAmount;
    uint256 public maliciousDkgResultSlashingAmountChangeInitiated;

    uint256 public newDkgResultChallengePeriodLength;
    uint256 public dkgResultChallengePeriodLengthChangeInitiated;

    uint256 public newRelayEntrySubmissionEligibilityDelay;
    uint256 public relayEntrySubmissionEligibilityDelayChangeInitiated;

    uint256 public newDkgSubmissionEligibilityDelay;
    uint256 public dkgSubmissionEligibilityDelayChangeInitiated;

    uint256 public newRelayEntryHardTimeout;
    uint256 public relayEntryHardTimeoutChangeInitiated;

    uint256 public newGroupCreationFrequency;
    uint256 public groupCreationFrequencyChangeInitiated;

    uint256 public newGroupLifetime;
    uint256 public groupLifetimeChangeInitiated;

    uint256 public newCallbackGasLimit;
    uint256 public callbackGasLimitChangeInitiated;

    // (...)
    // all `initialize*` and `finalize*` functions
    // (...)
}

// (owned by RandomBeaconGovernance)
contract RandmomBeacon is Ownable {
    /// @notice Relay request fee in T
    uint256 public relayRequestFee;

    /// @notice Reward for submitting DKG result
    uint256 public dkgResultSubmissionReward;

    /// @notice Reward for unlocking the sortition pool if DKG timed out
    uint256 public sortitionPoolUnlockingReward;

    /// @notice Slashing amount for not submitting relay entry
    uint256 public relayEntrySubmissionFailureSlashingAmount;

    /// @notice Slashing amount for submitting malicious DKG result
    uint256 public maliciousDkgResultSlashingAmount;

    /// @notice The number of blocks for which a DKG result can be challenged
    uint256 public dkgResultChallengePeriodLength;

    /// @notice The number of blocks for a member to become eligible to submit
    ///         relay entry
    uint256 public relayEntrySubmissionEligibilityDelay;

    /// @notice The number of blocks for a member to become eligible to submit
    ///         DKG result
    uint256 public dkgSubmissionEligibilityDelay;

    /// @notice Hard timeout for a relay entry
    uint256 public relayEntryHardTimeout;

    /// @notice The frequency of a new group creation
    uint256 public groupCreationFrequency;

    /// @notice Group lifetime
    uint256 public groupLifetime;

    /// @notice Callback gas limit
    uint256 public callbackGasLimit;

    function updateRelayEntryParameters(
        uint256 _relayRequestFee,
        uint256 _relayEntrySubmissionEligibilityDelay,
        uint256 _relayEntryHardTimeout,
        uint256 _relayEntryCallbackGasLimit
    ) external onlyOwner {
        // (...)
    }

    function updateGroupCreationParameters(
        uint256 _groupCreationFrequency,
        uint256 _groupLifetime,
        uint256 _dkgResultChallengePeriodLength,
        uint256 _dkgResultSubmissionEligibilityDelayx
    ) external onlyOwner {
        // (...)
    }

    function updateRewardParameters(
        uint256 _dkgResultSubmissionReward,
        uint256 _sortitionPoolUnlockingReward
    ) external onlyOwner {
        // (...)
    }

    function updateSlashingParameters(
        uint256 _relayEntrySubmissionFailureSlashingAmount,
        uint256 _maliciousDkgResultSlashingAmount
    ) external onlyOwner {
        // (...)
    }
        
}


```

tomaszslabon commented on solidity/random-beacon/hardhat.config.ts:
```diff
@@ -11,7 +11,7 @@ const config: HardhatUserConfig = {
+        version: "0.8.6",
       },
     ],
   },
```
Done in https://github.com/keep-network/keep-core/pull/2613/commits/f77e46a77c23e8dccc5d69b2bf3a7e1c72d4e417.

tomaszslabon commented on solidity/random-beacon/hardhat.config.ts:
```diff
@@ -11,7 +11,7 @@ const config: HardhatUserConfig = {
+        version: "0.8.6",
       },
     ],
   },
```
Done in https://github.com/keep-network/keep-core/pull/2613/commits/0e2630ed22d9d08521018437eb51929b24497c01.

nkuba commented on solidity/random-beacon/hardhat.config.ts:
```diff
@@ -6,12 +6,14 @@ import "@nomiclabs/hardhat-waffle"
       {
-        version: "0.8.5",
+        version: "0.8.6",
       },
```
What if we enable optimizer, will it help to reduce the contract size (https://github.com/keep-network/keep-core/pull/2630/commits/2e1ae8b28f68ef1c911f8e8cc4ae78f7a5ec1cbf)?

tomaszslabon replied:
It didn't reduce the contract size, but I added it anyway in https://github.com/keep-network/keep-core/pull/2613/commits/47170880fb7f15d26a42ab2f2575c7d4123ebf29, maybe it will generate more efficient code later.

nkuba commented on solidity/random-beacon/contracts/RandomBeaconParameters.sol:
```diff
@@ -0,0 +1,984 @@
+///         update
+/// @dev The client contract should take care of authorizations according to
+///      their needs.
+library GovernableParameters {
```
Don't we want to have the group size as a parameter?

pdyraga replied:
Not as a governable one according to the RFC.

nkuba commented:
I wanted to see if a more generic implementation based on storing properties in the storage slots (inspired by [eip-1967](https://eips.ethereum.org/EIPS/eip-1967)) would give better results. Please see: https://github.com/keep-network/keep-core/pull/2632
TLDR; It gives a much lower deployment cost, but a slightly higher get parameter cost.

tomaszslabon commented:
> I was experimenting with the contract size and I am afraid we still need to do better. The current version is more optimal than the previous one, but we are still above 9KB. I tried various options and the best I could come up with is to split all the governance-related functionalities into a separate contract, called `RandomBeaconGovernance` that would own `RandomBeacon` contract.
> 
> `RandomBeaconGovernance` would contain all `initialize*` and `finalize*` functions and we can even put there `getRemaining*` functions we originally had. `RandomBeacon` would expose four functions allowing the governance to update parameters, grouped by the parameter type (relay request, DKG, slashing, rewards). I'd also skip all the complexity of other types than `uint256`.
> 
> This should be pretty gas-optimal and the size of `RandomBeacon` contract would be just about 2.5KB.
> 
> What do you think about it?
> 
> Below is a sample code to help to visualize this proposition:
> 
> ```
> pragma solidity ^0.8.6;
> 
> import "@openzeppelin/contracts/access/Ownable.sol";
> 
> contract RandomBeaconGovernance is Ownable {
>     uint256 public newRelayRequestFee;
>     uint256 public relayRequestFeeChangeInitiated;
> 
>     uint256 public newDkgResultSubmissionReward;
>     uint256 public dkgResultSubmissionRewardChangeInitiated;
> 
>     uint256 public newSortitionPoolUnlockingReward;
>     uint256 public sortitionPoolUnlockingRewardChangeInitiated;
> 
>     uint256 public newRelayEntrySubmissionFailureSlashingAmount;
>     uint256 public relayEntrySubmissionFailureSlashingAmountChangeInitiated;
> 
>     uint256 public newMaliciousDkgResultSlashingAmount;
>     uint256 public maliciousDkgResultSlashingAmountChangeInitiated;
> 
>     uint256 public newDkgResultChallengePeriodLength;
>     uint256 public dkgResultChallengePeriodLengthChangeInitiated;
> 
>     uint256 public newRelayEntrySubmissionEligibilityDelay;
>     uint256 public relayEntrySubmissionEligibilityDelayChangeInitiated;
> 
>     uint256 public newDkgSubmissionEligibilityDelay;
>     uint256 public dkgSubmissionEligibilityDelayChangeInitiated;
> 
>     uint256 public newRelayEntryHardTimeout;
>     uint256 public relayEntryHardTimeoutChangeInitiated;
> 
>     uint256 public newGroupCreationFrequency;
>     uint256 public groupCreationFrequencyChangeInitiated;
> 
>     uint256 public newGroupLifetime;
>     uint256 public groupLifetimeChangeInitiated;
> 
>     uint256 public newCallbackGasLimit;
>     uint256 public callbackGasLimitChangeInitiated;
> 
>     // (...)
>     // all `initialize*` and `finalize*` functions
>     // (...)
> }
> 
> // (owned by RandomBeaconGovernance)
> contract RandmomBeacon is Ownable {
>     /// @notice Relay request fee in T
>     uint256 public relayRequestFee;
> 
>     /// @notice Reward for submitting DKG result
>     uint256 public dkgResultSubmissionReward;
> 
>     /// @notice Reward for unlocking the sortition pool if DKG timed out
>     uint256 public sortitionPoolUnlockingReward;
> 
>     /// @notice Slashing amount for not submitting relay entry
>     uint256 public relayEntrySubmissionFailureSlashingAmount;
> 
>     /// @notice Slashing amount for submitting malicious DKG result
>     uint256 public maliciousDkgResultSlashingAmount;
> 
>     /// @notice The number of blocks for which a DKG result can be challenged
>     uint256 public dkgResultChallengePeriodLength;
> 
>     /// @notice The number of blocks for a member to become eligible to submit
>     ///         relay entry
>     uint256 public relayEntrySubmissionEligibilityDelay;
> 
>     /// @notice The number of blocks for a member to become eligible to submit
>     ///         DKG result
>     uint256 public dkgSubmissionEligibilityDelay;
> 
>     /// @notice Hard timeout for a relay entry
>     uint256 public relayEntryHardTimeout;
> 
>     /// @notice The frequency of a new group creation
>     uint256 public groupCreationFrequency;
> 
>     /// @notice Group lifetime
>     uint256 public groupLifetime;
> 
>     /// @notice Callback gas limit
>     uint256 public callbackGasLimit;
> 
>     function updateRelayEntryParameters(
>         uint256 _relayRequestFee,
>         uint256 _relayEntrySubmissionEligibilityDelay,
>         uint256 _relayEntryHardTimeout,
>         uint256 _relayEntryCallbackGasLimit
>     ) external onlyOwner {
>         // (...)
>     }
> 
>     function updateGroupCreationParameters(
>         uint256 _groupCreationFrequency,
>         uint256 _groupLifetime,
>         uint256 _dkgResultChallengePeriodLength,
>         uint256 _dkgResultSubmissionEligibilityDelayx
>     ) external onlyOwner {
>         // (...)
>     }
> 
>     function updateRewardParameters(
>         uint256 _dkgResultSubmissionReward,
>         uint256 _sortitionPoolUnlockingReward
>     ) external onlyOwner {
>         // (...)
>     }
> 
>     function updateSlashingParameters(
>         uint256 _relayEntrySubmissionFailureSlashingAmount,
>         uint256 _maliciousDkgResultSlashingAmount
>     ) external onlyOwner {
>         // (...)
>     }
>         
> }
> ```

Done in https://github.com/keep-network/keep-core/pull/2613/commits/cb1e3b0498564d383c7aef304e495fbb1c6a8ae2.
`Random Beacon`'s size was reduced from `9.031 KB` to `2.909 KB`.

pdyraga commented on solidity/random-beacon/contracts/GovernanceUtils.sol:
```diff
@@ -0,0 +1,37 @@
+    /// @param changeTimestamp Timestamp indicating the beginning of the change.
+    /// @param delay Governance delay.
+    /// @return Remaining time in seconds.
+    function getRemainingChangeTime(uint256 changeTimestamp, uint256 delay)
```
I think `RandomBeaconGovernance` will be the only contract using this function. Can we make it `internal` in `RandomBeaconGovernance.sol`?

tomaszslabon replied:
Done in https://github.com/keep-network/keep-core/pull/2613/commits/082c206fef245c4b2095effd8eefce40622a5b25.

pdyraga commented on solidity/random-beacon/test/RandomBeacon.test.ts:
```diff
@@ -0,0 +1,213 @@
+  })
+
+  describe("updateRelayEntryParameters", () => {
+    const RELAY_REQUEST_FEE = 100
```
We were usually naming this kind of constants in camelCase. Is linter complaining about it?

tomaszslabon replied:
No, I could correct it without problems in https://github.com/keep-network/keep-core/pull/2613/commits/082c206fef245c4b2095effd8eefce40622a5b25.

pdyraga commented on solidity/random-beacon/contracts/RandomBeacon.sol:
```diff
@@ -0,0 +1,130 @@
+        groupCreationFrequency = _groupCreationFrequency;
+        groupLifetime = _groupLifetime;
+        dkgResultChallengePeriodLength = _dkgResultChallengePeriodLength;
+        dkgSubmissionEligibilityDelay = _dkgSubmissionEligibilityDelay;
```
It would be good to emit `GroupCreationParametersUpdated` event. Events are emitted by the governance contract but it would be good to treat this contract as a separate unit and have it maintain its changelog independently. Same for all other `update*` functions in `RandomBeacon.sol`.

tomaszslabon replied:
Done in https://github.com/keep-network/keep-core/pull/2613/commits/535aa4e6b5539d2cc9e24ccd27c9095522d890c2.

pdyraga commented on solidity/random-beacon/contracts/RandomBeaconGovernance.sol:
```diff
@@ -0,0 +1,841 @@
+    function beginRelayEntrySubmissionEligibilityDelayUpdate(
+        uint256 _newRelayEntrySubmissionEligibilityDelay
+    ) external onlyOwner {
+        /* solhint-disable not-rely-on-time */
```
We could consider silencing this rule in solhint, it is annoying.

tomaszslabon replied:
I agree, but I think we should wait till we have `solhint` enabled in `solidity/random-beacon`.
I added these  `/* solhint-disable not-rely-on-time */`statements because I thought they would be necessary.

pdyraga replied:
Agreed, I should be more clear - it's not something we should work on in this PR. I even think we should consider ignoring it in the Thesis-level solhint config.

pdyraga reviewed (approved):
Looks good, I have some non-critical, non-blocking comments but I'll address them in a separate PR instead of complaining here.
